include/jemalloc/internal/tcache_inlines.h

```diff
@@ -33,68 +33,80 @@ tcache_enabled_set(tsd_t *tsd, bool enabled) {
 static inline size_t
-tcache_max_get(tcache_t *tcache) {
-	assert(tcache != NULL);
-	assert(tcache->tcache_max <= TCACHE_MAXCLASS_LIMIT);
-	return tcache->tcache_max;
+tcache_max_get(tcache_slow_t *tcache_slow) {
+	assert(tcache_slow != NULL);
+	size_t tcache_max = sz_index2size(tcache_nbins_get(tcache_slow) - 1);
+	assert(tcache_max <= TCACHE_MAXCLASS_LIMIT);
+	return tcache_max;
 }
 
 static inline void
-tcache_max_and_nhbins_set(tcache_t *tcache, size_t tcache_max) {
-	assert(tcache != NULL);
+tcache_max_set(tcache_slow_t *tcache_slow, size_t tcache_max) {
+	assert(tcache_slow != NULL);
 	assert(tcache_max <= TCACHE_MAXCLASS_LIMIT);
-	tcache->tcache_max = tcache_max;
-	tcache->tcache_nhbins = sz_size2index(tcache_max) + 1;
+	tcache_slow->tcache_nbins = sz_size2index(tcache_max) + 1;
+}
+
+static inline void
+tcache_bin_settings_backup(tcache_t *tcache,
+    cache_bin_info_t *tcache_bin_info) {
+	for(unsigned i = 0; i < TCACHE_NBINS_MAX; i++) {
+		cache_bin_info_init(&tcache_bin_info[i],
+		    tcache->bins[i].bin_info.ncached_max);
+	}
 }
 
 static inline void
-thread_tcache_max_and_nhbins_set(tsd_t *tsd, size_t tcache_max) {
+thread_tcache_max_set(tsd_t *tsd, size_t tcache_max) {
 	assert(tcache_max <= TCACHE_MAXCLASS_LIMIT);
 	assert(tcache_max == sz_s2u(tcache_max));
 	tcache_t *tcache = tsd_tcachep_get(tsd);
-	tcache_slow_t *tcache_slow;
-	assert(tcache != NULL);
+	tcache_slow_t *tcache_slow = tcache->tcache_slow;
+	cache_bin_info_t tcache_bin_info[TCACHE_NBINS_MAX] = {{0}};
+	assert(tcache != NULL && tcache_slow != NULL);
 
 	bool enabled = tcache_available(tsd);
 	arena_t *assigned_arena;
 	if (enabled) {
-		tcache_slow = tcache_slow_get(tsd);
-		assert(tcache != NULL && tcache_slow != NULL);
 		assigned_arena = tcache_slow->arena;
+		/* Carry over the bin settings during the reboot. */
+		tcache_bin_settings_backup(tcache, tcache_bin_info);
 		/* Shutdown and reboot the tcache for a clean slate. */
 		tcache_cleanup(tsd);
 	}
 
 	/*
-	* Still set tcache_max and tcache_nhbins of the tcache even if
-	* the tcache is not available yet because the values are
-	* stored in tsd_t and are always available for changing.
+	* Still set tcache_nbins of the tcache even if the tcache is not
+	* available yet because the values are stored in tsd_t and are
+	* always available for changing.
 	*/
-	tcache_max_and_nhbins_set(tcache, tcache_max);
+	tcache_max_set(tcache_slow, tcache_max);
 
 	if (enabled) {
-		tsd_tcache_data_init(tsd, assigned_arena);
+		tsd_tcache_data_init_with_bin_settings(tsd, assigned_arena,
+		    tcache_bin_info);
 	}
 
-	assert(tcache_nhbins_get(tcache) == sz_size2index(tcache_max) + 1);
+	assert(tcache_nbins_get(tcache_slow) == sz_size2index(tcache_max) + 1);
 }
 
 JEMALLOC_ALWAYS_INLINE bool
 tcache_small_bin_disabled(szind_t ind, cache_bin_t *bin) {
 	assert(ind < SC_NBINS);
 	assert(bin != NULL);
-	bool ret = cache_bin_info_ncached_max(&bin->bin_info) == 0;
+	bool ret = cache_bin_info_ncached_max(bin, &bin->bin_info) == 0;
```

Comment:
Now with ncached_max getter also wrap around another logic inside, this bin_disabled logic is getting out of control.

Let's drill this down and only have one consistent way of determining enabled or not: that is the `cache_bin_disabled` which does the `stack_head == cache_bin_disabled_stack()` check.

We need to completely disconnect ncached_max and is_disabled checks -- since we are saying that users may have ncached_max values set to non zero but disable them using tcache_max. The cache_bin_disabled check is straightforward and not layered, while the current approach, even if we get it correct right now, is too error prone for future changes.

That being said, it means we should init large bins the same way too. So that all bins are setup with the below properties:
1) if disabled, stack_head points to disabled_bins, ncached_max can be 0 or not.
2) if enabled, stack_head does not point to disabled_bins, with ncached_max > 0
2) if ncached_max is zero, must be disabled
3) if ncached_max is non zero, can only be disabled when it's bin index > tcache_nbins